casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -8,84 +8,217 @@ import cats.{Applicative, Monad}
+          ts <- Time[F].currentMillis
+        } yield ((hash -> Option(modifiedRequested(ts))))
+      } else {
+        val warnMessage = s"Could not retrieve requested block ${PrettyPrinter.buildString(hash)}. " +
+          "Removing the request from the requested blocks list. Casper will have to re-request the block."
+        Log[F].warn(warnMessage).as((hash -> none[Requested]))
+      }
+
+    RequestedBlocks[F].flatModify(requests => {
+      requests.keys.toList
+        .traverse(hash => {
+          val requested = requests(hash)
+          Time[F].currentMillis
+            .map(_ - requested.timestamp > timeout.toMillis)
+            .ifM(tryRerequest(hash, requested), (hash -> Option(requested)).pure[F])
+
+        })
+        .map(list => toMap(list))
+    })
+  }
+
+  def handleHasBlock[F[_]: Monad: RPConfAsk: RequestedBlocks: TransportLayer: Time: Log](
+      peer: PeerNode,
+      hb: HasBlock
+  )(
+      casperContains: BlockHash => F[Boolean]
+  ): F[Unit] = {
+
+    val hashStr = PrettyPrinter.buildString(hb.hash)
+
+    def addNewEntry =
+      Log[F].info(s"Creating new entry for the $hashStr request") >> Time[F].currentMillis >>= (
+          ts => {
+            RequestedBlocks[F].modify(
+              _ + (hb.hash -> Requested(timestamp = ts, peers = Set(peer)))
+            )
+          }
+      )
+
+    def addToWaitingList(requested: Requested): F[Unit] =
+      Log[F].info(s"Adding $peer to waiting list of $hashStr request") >> RequestedBlocks[F]
+        .modify { requestedBlocks =>
+          requestedBlocks + (hb.hash -> requested.copy(waitingList = peer :: requested.waitingList))
+        }
+
+    val logIntro = s"Received confirmation from $peer that it has block $hashStr."
+    Log[F].info(logIntro) >> casperContains(hb.hash).ifM(
+      noop[F],
+      RequestedBlocks[F].read >>= (_.get(hb.hash)
+        .fold(requestForBlock[F](peer, hb.hash) >> addNewEntry)(addToWaitingList))
+    )
+  }
+
+  def handleHasBlockRequest[F[_]: Monad: RPConfAsk: TransportLayer](
+      peer: PeerNode,
+      hbr: HasBlockRequest
+  )(blockLookup: BlockHash => F[Boolean]): F[Unit] =
+    blockLookup(hbr.hash).ifM(
+      for {
+        conf <- RPConfAsk[F].ask
+        msg = packet(
+          conf.local,
+          conf.networkId,
+          transport.HasBlock,
+          HasBlock(hbr.hash).toByteString
+        )
+        _ <- TransportLayer[F].send(peer, msg)
+      } yield (),
+      noop[F]
+    )
+
+  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks](peer: PeerNode, b: BlockMessage)(
+      casperContains: BlockHash => F[Boolean],
+      casperAdd: BlockMessage => F[BlockStatus]
+  ): F[Unit] =
+    casperContains(b.blockHash)
+      .ifM(
+        Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again."),
+        for {
+          _ <- Log[F].info(s"Received ${PrettyPrinter.buildString(b)}.")
+          _ <- casperAdd(b)
+          _ <- RequestedBlocks[F].modify(_ - b.blockHash)
```

Comment:
We should add a TODO or a ticket to check BlockStatus for inDag and only remove from RequestedBlocks if so

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3675